### DISCUSSION TITLE
Value dependant types

### DISCUSSION THREAD

**Discussion by chrjohannesen** [sinclairzx81/typebox#270] [Q&A] [answered]
Hey!

Just found this gem of a repo, and thought I'd give a whirl. It's working incredibly well and could potentially save a lot of work! Nice job  🎉

I've run into a headscratcher as I can't seem to find out how to solve the following TS types with Typebox. I've built a query builder of sort which means the JSON object I can pass through have different variations.

As an example:
```ts
// The type of value changes dependant on the operator value
const body = {
	set: 'dog',
	filters: [
		{
			field: 'breed',
			operator: 'in',
			value: ['Beagle', 'Labrador']
		},
		{
			field: 'weight',
			operator: 'range',
			value: {
				from: 0,
				to: 100
			}
		},
	]
};
```

Another variation of the same issue could be this, which can be cumbersomely handled by splitting the validation.

```ts
import Ajv from 'ajv'
import { Static, Type } from "@sinclair/typebox";

const ajv = new Ajv();

function StringEnum<T extends string[]>(values: [...T]) {
	return Type.Unsafe<T[number]>({ type: 'string', enum: values })
}

const ImportSchema = Type.Object({
	importIds: Type.Array(Type.Integer())
})

const UserSchema = Type.Object({
	userIds: Type.Array(Type.Integer())
})

const DefinitionSchemas = {
	import: ImportSchema,
	user: UserSchema,
}

const ConfigSchema = Type.Object({
	campaignId: Type.Number(),
	type: StringEnum(['import', 'user']),
	definition: Type.Unknown()
})


const segment: Static<typeof ConfigSchema> = {
	"type": "user",
	"campaignId": 1,
	"definition": {
		"userIds": [1, 2, 3]
	}
}

if (
	ajv.validate(ConfigSchema, segment)
	&& ajv.validate(DefinitionSchemas[segment.type], segment.definition)
) {
	console.log('Hooraaay!')
}

console.log(ajv.errors)
```

So my question really is, how do I create a type that changes dependant on a value of a sibling with validation for the TypeCompiler or ajv?

Hope to hear from anyone smarter than me! 🤞🏼

**Comment by chrjohannesen** [comment, marked as answer]
I've also tried something like this, which don't work: 
```ts
const ImportSchema = Type.Object({
	type: Type.Literal('import'),
	definition: Type.Object({
		importIds: Type.Array(Type.Integer())
	})
})

const UserSchema = Type.Object({
	type: Type.Literal('import'),
	definition: Type.Object({
		userIds: Type.Array(Type.Integer())
	})
})

export const ConfigSchema = Type.Intersect([
	Type.Object({
		campaignId: Type.Number(),
	}),
	Type.Union([ImportSchema, UserSchema])
])
```

**Reply by sinclairzx81** [reply]
Hi, the following should produce a schema that closely aligns the `filters` / `operators` example in your original question.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxnAL5wBmUEIcA5AAIDOwAdiQDYHBQD0MdBgBGEAB4MAsAChZ-fnAC0K1WvUbNW7evmKAYsC4wMUOAGEI7ACbBiVlnuU6Xrt6tlOqRk1AIiuDDgANwIuAFcMFjgSCHCuazgRIJYYKA4AczgAHzh2cJBks1yxCECCdlkMcUhYGId4ADUwyLgAXhRhADoAVXZgKwAKAG1OzC78dPYMwYBKHDRxgDkCormF7oAhCDKMCrm4AF1ZzxkFeps7AfYALjhB0IiMLrBaMHa2ju9jUy7HyJOMmqtXgsXYqTgAFEAI7hMLRDqLZ4AeREACsMCQYIMELI4HAIJg-DBoHckV0ADJ2UxhQYMDCw+EMeZ46jADAJMndSaZdas-4YO7NJ6yciApxg2z2W73b6+P4tZ4cbjhaxRB6Kl5vWaA4HQUENOAASXY7TGKPRmOxuJk+MJNJJUC54ypvlpDA4zKwrO8HOszuePOmfNtIUVAa6AEEoH5UINhQDReKzopJVcrHcNU8tYS4AA+L4+X40OhwABkZbD2deuYAPGyflAuiTdTV9RcIQAlCoZIKI7qojFYnGs+3E0nmynUvxcOl+aYYL0+9mcydBmYs0MCiODq0j0P4ksgCMrQqmEP4-Ekk+rc+A-FipOnPV1MEQwyNs3kvrXEYwuFcCwOAmjg3YLscpznO40EwboKZwD0YDWEQGBOLB6HQc+bavoaiHISYX4Dpaw42viLAYDAEbrheDa+CwEbRrGgwfr4gKPnI8EYVxLhOMgUQwGh3FCRokGKCS1gQHcoBgIEIAYOwBqXNKYRwOESEoXAXAQBkpCyFQ4ScNKqnqSYgxqfhgp4IQxAkLWQiYBAVAISZGB5rMiA+tAgxvvAcqmASTnmShXR+VALDuaRl4sAA7nYJAABaDKFXRjkQ0ARayl4xAQ5GMAyAEsAwdz8AAVCVcAlYoIhQHsADWmWXiQOVBB67BFXApXlZVSQ1QQ9UHo1zWMPOvbtZ1FVVb1-VZeQrKzTI82yEFpmReRlGMBJGQMN6oahfRcDDA1kVZb6q4MNVGAYNY20NXaRJpU6jCejtWX4tuB0MJsewZIE22MBS-h+BJUAMIcDXkC9l7HZep3+ow0UYMAGTxTAN0DXdDoTgwI2LpDWXvdDJ20MecAAAx46915wAAjKTpMdam8U9kEJJwKkUxZFQ0BwKYtBQLdFDg5DhxAA)

```typescript
import { Type, Static } from '@sinclair/typebox'

// --------------------------------------------------
// Filter Conditions
// --------------------------------------------------

// filterable values could be string | number | boolean
export const Value = Type.Union([ Type.String(), Type.Number(), Type.Boolean() ])

// condition: (value.prop === filter.value)
export const Equals = Type.Object({
  operator: Type.Literal('equals'),
  field: Type.String(),
  value: Value
})

// condition: (filter.value.includes(value.prop))
export const In = Type.Object({
  operator: Type.Literal('in'),
  field: Type.String(),
  value: Type.Array(Value)
})

// condition: (value.prop >= filter.from && value.prop < filter.to)
export const Range = Type.Object({
  operator: Type.Literal('range'),
  field: Type.String(),
  value: Type.Object({
    from: Type.Number(),
    to: Type.Number()
  })
})

export const Filter = Type.Union([Equals, In, Range])

// --------------------------------------------------
// Update
// --------------------------------------------------

export const Update = Type.Object({
  set: Type.String(),
  filters: Type.Array(Filter)
})

// --------------------------------------------------
// Test
// --------------------------------------------------

// todo: implement conditional update logic
function update(update: Static<typeof Update>) {
  for(const filter of update.filters) {
    switch(filter.operator) {
      case 'equals': /** */ break
      case 'in': /** */ break
      case 'range': /** */ break
    }
  }
}

update({
  set: 'dog',
  filters: [
    {
      field: 'breed',
      operator: 'in',
      value: ['Beagle', 'Labrador']
    },
    {
      field: 'weight',
      operator: 'range',
      value: {
        from: 0,
        to: 100 // change to string for error
      }
    },
  ]
})
```

> The `Type.Union()` maps to `type T = A | B | C` and `Type.Intersect()` maps to `type T = A & B & C`.

It should be possible to constrain this schema more (for example, only allowing a filter `field` to be set to known properties of the type being filtered (for example, only allowing `field` to be set to the properties of the `dog` type in your example)). To achieve this, take a look at [Generic Types](https://github.com/sinclairzx81/typebox#types-generic) and the `Type.KeyOf()` function.

Hope this helps!

**Reply by chrjohannesen** [reply]
Hey @sinclairzx81!

Thank you so much for taking the time to compose this example. It's super neat! That actually fixes the issues I had on that part. As you also state, there's gonna be some validation errors that are not applicable in certain scenarios, as the Union type will not discard whatever filter is not in "use". I'll have a crack at that, but I think it's kinda the same issue I have here:

```ts
const ImportSchema = Type.Object({
	importIds: Type.Array(Type.Integer())
})

const UserSchema = Type.Object({
	userIds: Type.Array(Type.Integer())
})

const Definitions = Type.Object({
	import: ImportSchema,
	user: UserSchema
})

const ConfigSchema = Type.Object({
	type: Type.KeyOf(Definitions),
	campaignId: Type.Number(),
	definition: Type.Pick(Definitions, Type.KeyOf(Definitions))
});
```

Again the goal is to only have the definition given by the key ("import" or "user"). I think I'm treading a level of TS types and interfaces out of my league for now.

Nonetheless thank you so much for your time! 🎉

**Reply by chrjohannesen** [reply]
It's essentially this I want:
```ts
interface UserDefinition {
	userIds: Id[];
}

interface UserSegmentation {
	type: 'user';
	definition: UserDefinition
}

interface ImportDefinition {
	importIds: Id[]
}

interface ImportSegmentation {
	type: 'import',
	definition: ImportDefinition
}

type SegmentationContrains = UserSegmentation | ImportSegmentation;

interface Segmentation<T extends SegmentationContrains> {
	type: T['type'];
	referenceId: Id;
	campaignId: Id;
	definition: T['definition']
}
```

**Reply by sinclairzx81** [reply]
@chrjohannesen Hey no problem :)

I'd recommend maybe trying to implement your structures with TypeScript first and use that as a reference to guide you through the creation of the corresponding TypeBox type. In both examples, I think generics may be the most idealized way to express a common structure that is dependent on a varying type (relevant both to `filters` and to `configs`). 

So, just to revisit the previous filter, here would be the TypeScript only generics approach to express that.

### TypeScript

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/PTAEFpK6dv4Y6BYAUCUAVAngBwKYDKAxgE4CWuALmhkvQ4wmlXvqAGoCGANgK7sAvKADOVCgDsA5qAA+oCXwC2AI3yk5oFQHttPfFwlo0+AB65tpKqHISq6gGZdi7AKIBHPrxEAeTAD5QAG80UDDtAlIuKksALlAAcg8vHhEEgBpQsIBrfGx43OxtByxM1DDQADdeAXjufnw0AF8Tc0trW3tSJxdQAEkJP0CQ8vDI6LjEgYys0EKCvOLS2eqGupr8AG0AXWbWiytQVgJQAHlxmNIh0GFk72v5AaHjVDMDjrtHZ3YAMXIeLrXEYVBz-LoieLndQTK4BHZ7VC0MBMFGopBI0CuUxcJS4fQYtGEomQF6dL69Dj4YiXYJZUzxRSqdRlML5BTKNSkFmgABeDI56gRaDJ3W+oAAwnwxNolOpaaNyAATeJiSRSLISHH4FXiWxSbn4JRcf46tVC1DHdiU6mWP4AuVhYR2wHWy7+FhsCVSmKy0jOuVOsHqHyS6W+92IyMOPgSankbQSKpUy7+0gAClWtU4ydtQdIAEpgi1I5Uc3682ngaBQfbSBDQJtZsFQBFoZd4klPN4MnM8h3TD3M9rQABGUBNblhIIti6TBLTdK9tkJbCDjbxTYjxcAJkXAGZtuOsrtUE18y9o7GqPHE8RvTL1KmM+uvWHH3nC0Fi2g72-y7XKyyGtwQ3Jtp1bKJ20SO5Uh7eZEiVNc1lHcdJ2bCCYQ7Bclw7TVZSQrNNgSRUuFLHsEhEJRyCoAALBJD2LMITzPIA)

```typescript
// -----------------------------------------------------------
// TypeScript
// -----------------------------------------------------------
type Value = string | number | boolean

export interface Equals<T> {
   operator: 'Equals',
   key: keyof T,
   value: Value
}
export interface In<T> {
   operator: 'In',
   key: keyof T,
   value: Value[]
}
export type Operator<T> = Equals<T> | In<T>

export interface Filter<T> {
   filters: Operator<T>[]
}

// -----------------------------------------------------------
// Example
// -----------------------------------------------------------

interface Vector {
  x: number,
  y: number,
  z: number
}

interface Customer {
  id: string
  name: string,
  email: string
}

type VectorFilter   = Filter<Vector>
type CustomerFilter = Filter<Customer>


function vectorFilter(value: VectorFilter) {}

vectorFilter({
  filters: [
    { operator: 'Equals', key: 'x', value: 1 },
    { operator: 'In', key: 'y', value: [1, 2, 3] }
  ]
})

function customerFilter(value: CustomerFilter) {}

customerFilter({
  filters: [
    { operator: 'Equals', key: 'id', value: 1 },
    { operator: 'In', key: 'name', value: ['dave', 'smith'] }
  ]
})
```

### TypeBox

So once you have the TypeScript types working, representing that in TypeBox should be a lot easier. The following implements the above using TypeBox.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxjsgPIBGAVhifAL5wBmUEIcA5AAIDOwAHYkANgWBQA9DHQYqEAB5cAsACg1kyXAC0uvfoOGjxk6f0ataTACFFFnWcdPnxtWowLIsOCQiC+8ABqBCIArhhwALwosgB0AKqCwH4AFADaVhix+FBCAOYpAJTkcQByoSBUGFBFJZixthAiGASCRQC6hW6qHl7wvv7wAKIAjqEhfFFwADzIcB4wGIIAJpOUtPQwAHwpfCQAFhggBABcKIVRWzH11HQMKQhqcHAQmFBE0GeZsQAywIvvEQpLijcYiPhcYpPOAAawwqC+cQA0vCKCxdgcjgQoapngA3ELhM7BMIYNSMLrqHqeaD9PwBOAASUEU1m8wUixWa1umx2e0Oxy+F0iV2+PPuj1xLzeHygiPqfwBIWBzMhWGhcIR1yyKNQaIxAux6qlBNJ8qyAEEoO9UCkSeFKRT3DTvAMGRQZTBoKy5gslqsUOLtgasULLtqEklUmloaCJiHjjjnsyE9i1J1ur1aT56fAAGLAEQAn3szkB9Z3YP80PncNijYS6EsQsAvjm2JWm0pD3VWWpwqOyndTQOFxj8cmexDBQEcDNewTxdL3TdN1BTbe6L1ysPaEKdvlSrVWrQrXfQ9VGpJuAALwPFUvRXJlLXcAAwqEAhxqlNt5td1KwDLO2OT5CeUqCLOGAgTAuSCAU15YoWMFwQhz6Zi6dKDHAgQblABZFj+zzRARAJ2nhlJZq6ubvp+XogNUpE-iRLbHh+X4MVAQ6qCwoTCMQfhwHieFMTUppEnghDECQ0wyJgEAsDhImsVAWwXAgjDdMJDDQKJAHPM2hFQG2cAxlKzxIK8vZenK3BxuCXA4JqZxcEoODidBcAAIxwIwxrPBZ0rWZ83Cqk58IuagjlCYSnlpF5OAAEw4AAzO0vnQu06FUrx-HJCyJB0d++EqSkHlnPgRCkLJsgKbRHGMSpamIJpVKFQ1JVGfprAqSZZkBYgQXvDZLn2RC4ValwQHReV3m+f5AWWZ6IVcGFsIRdwkEMTNsVnGkXDLAQwnRVwfAgP8+xcOlrXPFlqgUkAA)

```typescript
import { Type, Static, TObject } from '@sinclair/typebox'

// -----------------------------------------------------------
// TypeBox
// -----------------------------------------------------------

export const Value = Type.Union([Type.String(), Type.Number(), Type.Boolean()])

export const Equals = <T extends TObject>(schema: T) => Type.Object({
  operator: Type.Literal('Equals'),
  key: Type.KeyOf(schema),
  value: Value
})

export const In = <T extends TObject>(schema: T) => Type.Object({
  operator: Type.Literal('In'),
  key: Type.KeyOf(schema),
  value: Type.Array(Value)
})
export const Operator = <T extends TObject>(schema: T) => Type.Union([
  Equals(schema),
  In(schema)
])

export const Filter = <T extends TObject>(schema: T) => Type.Object({
  filters: Type.Array(Operator(schema))
})


// -----------------------------------------------------------
// Example
// -----------------------------------------------------------

const Vector = Type.Object({
  x: Type.Number(),
  y: Type.Number(),
  z: Type.Number()
})
const Customer = Type.Object({
  id: Type.String(),
  name: Type.String(),
  email: Type.String()
})

export const VectorFilter   = Filter(Vector)
export const CustomerFilter = Filter(Customer)

function vectorFilter(value: Static<typeof VectorFilter>) {}

vectorFilter({
  filters: [
    { operator: 'Equals', key: 'x', value: 1 },
    { operator: 'In', key: 'y', value: [1, 2, 3] }
  ]
})

function customerFilter(value: Static<typeof CustomerFilter>) {}

customerFilter({
  filters: [
    { operator: 'Equals', key: 'id', value: 1 },
    { operator: 'In', key: 'name', value: ['dave', 'smith'] }
  ]
})
```

I am actually hoping to provide [some tools](https://github.com/sinclairzx81/typebox/blob/master/codegen/typebox.ts#L31) around this to enable automatic TS > TB translation (as I do think it's generally easier to reason about complex generic types using familiar TypeScript syntax). I am hoping to provide code translation via an official website one day (likely when TypeBox reaches 1.0), but for now, manually constructing TB types is the only option (which can be quite complex)

Hope this helps

**Reply by chrjohannesen** [reply]
That's the approach I'm at right now and it's looking good, but I couldn't wrap my head around the generic types in TypeBox albeit your examples are super helpful. So this is super helpful and definitely answers the question about generic types.

I'll give it another go later. Thanks @sinclairzx81! 👍

**Reply by chrjohannesen** [reply]
I can't mark your response as the answer, so I've marked the thread!
